### DISCUSSION TITLE
How do I use mockall in a struct that derives Clone?

### DISCUSSION THREAD

**Discussion by augustocdias** [asomers/mockall#308] [Questions] [answered]
So I have a struct like this

```rust
#[derive(Clone)]
pub struct MyStruct {
...
}

#[automock]
impl MyStruct{
}
```

I use the `double` macro in the imports to not conflict the namespacing. But my tests fail to compile because it says the `MockMyStruct` crate does not implement `Clone`. Is there a way to achieve what I'm trying here?

And the function I wanna test at some point calls `clone` on this object and passes to another function. Is there a way to verify calls in this "cloned" object?

**Comment by asomers** [comment, marked as answer]
The way you specified it, Mockall doesn't know that `MyStruct` is supposed to be `Clone`.  Instead, do it like this:
```rust
mock! {
    MyStruct {}
    impl Clone for MyStruct {
        fn clone(&self) -> Self;
    }
}
```

**Reply by augustocdias** [reply]
I tried that, but then I have to add every method from my struct into the `mock!` and the `double` stops working so I can't pass this new mock to my function

**Reply by asomers** [reply]
`#[double]` should still work, if you structure your modules correctly.

**Reply by augustocdias** [reply]
So I have in the structs module the following 

```rust
#[derive(Clone)]
pub struct MyStruct {
}

#[automock]
impl MyStruct{
   pub fn my_func() {}
}
```

Then I have a function in another module:

```rust
#[double]
use my_mod::MyStruct;

fn function_i_wanna_test(my_struct: MyStruct) -> Result<(), Error> {}
```

I have a separate test module where I write my unit tests

```rust
use my_mod::MyStructMock;

mock !{
   MyStruct {
         pub fn my_func();
   }
   impl Clone for MyStruct {
        fn clone(&self) -> Self;
    }
}

#[test]
fn test_my_struct() {
  let mock = MyStructMock::default();
  function_i_wanna_test(mock);
}
```

This doesn't compile... How should I change then to make it work?

**Reply by asomers** [reply]
What is `MyStructMock`?  I don't see a definition for that anywhere.  And why are you using both `automock` and `mock!`?  You should use only one.  Do it like this:
```rust
pub mod my_mod {
    #[derive(Clone)]
    pub struct MyStruct {
    }

    impl MyStruct{
       pub fn my_func() {}
    }
    mock !{
       MyStruct {
             pub fn my_func();
       }
       impl Clone for MyStruct {
            fn clone(&self) -> Self;
        }
    }
}
mod test {
    #[double]
    use super::my_mod::MyStruct;
    ...
}
```

**Reply by augustocdias** [reply]
Sorry, it is `MockMyStruct`. It is the name generated by `mockall`. 

I was not using the `#[double]` in the tests. Which I'll start to use because it looks better :) 

Your suggestion works. Thanks for the help
